lib/workload/stateless/stacks/filemanager/filemanager-build/src/error.rs

```diff
@@ -0,0 +1,95 @@
+//! This module contains the crate's error types.
+//!
+
+use crate::error::ErrorKind::{IoError, LoadingEnvironment};
+use crate::workspace_path;
+use miette::{diagnostic, Diagnostic, NamedSource, SourceOffset};
+use std::fmt::{Display, Formatter};
+use std::fs::read_to_string;
+use std::panic::Location;
+use std::{io, result};
+use thiserror::Error;
+
+pub type Result<T> = result::Result<T, Error>;
+
+/// Error types for the filemanager.
+#[derive(Error, Debug)]
+pub enum ErrorKind {
+    #[error("Error generating entities: {0}")]
+    EntityGeneration(String),
+    #[error("Missing or incorrect environment variables: {0}")]
+    LoadingEnvironment(String),
+    #[error("io error: {0}")]
+    IoError(io::Error),
+}
+
+#[derive(Error, Debug, Diagnostic)]
+#[diagnostic(help("Is a local database instance running with environment variables set?"))]
+pub struct Error {
+    error_kind: ErrorKind,
+    #[source_code]
+    src: Option<NamedSource<String>>,
+    #[label]
+    label: Option<SourceOffset>,
+}
+
+impl Display for Error {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}", self.error_kind)
+    }
+}
+
+impl From<envy::Error> for Error {
+    #[track_caller]
+    fn from(error: envy::Error) -> Self {
+        Self::from(LoadingEnvironment(error.to_string()))
+    }
+}
+
+impl From<io::Error> for Error {
+    #[track_caller]
+    fn from(error: io::Error) -> Self {
+        Self::from(IoError(error))
+    }
+}
+
+impl From<ErrorKind> for Error {
+    /// Create an error with caller location to print error information.
+    #[track_caller]
+    fn from(error_kind: ErrorKind) -> Self {
+        let loc = Location::caller();
+
+        if let Some(path) = workspace_path() {
+            if let Ok(source) = read_to_string(path.join(loc.file())) {
+                let offset = SourceOffset::from_location(
+                    source.as_str(),
+                    loc.line() as usize,
+                    loc.column() as usize,
+                );
+
+                return Self::new(
+                    error_kind,
+                    Some(NamedSource::new(loc.file(), source)),
+                    Some(offset),
+                );
+            }
+        }
+
+        Self::new(error_kind, None, None)
+    }
```

Comment:
Using miette for nicer compile/build error printing. `#[track_caller]` is useful to print line information about where an error occurred, similar to the Rust compiler.